File: Assets/_Scripts/Views/PlayerView.cs
using RageCure.StateUtils;
using Styx.Commands;
using Styx.Entities;
using Styx.Entities.PlayerModule;
using Styx.Managers;
using Styx.Models;
using Styx.States;
using Styx.States.playerState;
using UnityEngine;

namespace Styx.Views
{
    public class PlayerView : MonoBehaviour
    {
        [SerializeField]
        private PlayerId _playerId;

        [SerializeField]
        private RigidbodyModel _rigidbodyModel;
        [SerializeField]
        private MovementModel _movementModel;
        [SerializeField]
        private AnimatorModel _animatorModel;
        [SerializeField]
        private TransformModel _transformModel;
        [SerializeField]
        private DashModel _dashModel;
        [SerializeField]
        private RagdollModel _ragdollModel;
        [SerializeField]
        private SlapshotModel _slapshotModel;

        [SerializeField]
        private Rigidbody _rigidbody;

        // *********************** INPUTS ***********************************
        [SerializeField]
        private InputAxisCommand _horizontalAxis;
        [SerializeField]
        private InputAxisCommand _verticalAxis;
        [SerializeField]
        private InputAxisCommand _horizontalShootAxis;
        [SerializeField]
        private InputAxisCommand _verticalShootAxis;
        [SerializeField]
        private InputKeyDownCommand _passKeyDown;
        [SerializeField]
        private InputKeyDownCommand _dashKeyDown;
        // *******************************************************************

        // *************************** STATES ******************************************
        [SerializeField]
        private PlayerCarryPuckState _playerCarryPuckState;
        [SerializeField]
        private PlayerNormalState _playerNormalState;
        [SerializeField]
        private PlayerPassState _playerPassState;
        [SerializeField]
        private PlayerDashState _playerDashState;
        [SerializeField]
        private PlayerKnockedDownState _playerKnockedDownState;
        [SerializeField]
        private PlayerSlapShotState _playerSlapShotState;

        public PlayerState PlayerState { get; private set; }

        public void Awake()
        {
            _playerId = (_playerId ?? GetComponent<PlayerId>());

            _playerCarryPuckState = (_playerCarryPuckState ?? GetComponent<PlayerCarryPuckState>());
            _playerNormalState = (_playerNormalState ?? GetComponent<PlayerNormalState>());
            _playerPassState = (_playerPassState ?? GetComponent<PlayerPassState>());
            _playerDashState = (_playerDashState ?? GetComponent<PlayerDashState>());
            _playerKnockedDownState = (_playerKnockedDownState ?? GetComponent<PlayerKnockedDownState>());
            _playerSlapShotState = (_playerSlapShotState ?? GetComponent<PlayerSlapShotState>());
        }
        
        public void Start()
        {
            var animator = GetComponent<Animator>();
            _rigidbody = _rigidbody ?? GetComponent<Rigidbody>();
            _transformModel.Transform = transform;
            _animatorModel.Animator = animator;
            _ragdollModel.MainAnimator = animator;
            _ragdollModel.MainCollider = GetComponent<CapsuleCollider>();
            _ragdollModel.MainRigidbody = _rigidbody;

            //TODO: Will be moved when team management system and game loop system is implemented
            //TODO: Use a builder pattern maybe??
            PlayerState = new PlayerState
            {
                Player = new Player
                {
                    Id = _playerId,
                    RigidbodyModel = _rigidbodyModel,
                    MovementModel = _movementModel,
                    TransformModel = _transformModel,
                    AnimatorModel = _animatorModel,
                    DashModel = _dashModel,
                    RagdollModel = _ragdollModel,
                    SlapshotModel = _slapshotModel,
                    Rigidbody = _rigidbody,
                    Inputs = new Inputs
                    {
                        HorizontalAxis = _horizontalAxis,
                        VerticalAxis = _verticalAxis,
                        HorizontalShootAxis = _horizontalShootAxis,
                        VerticalShootAxis = _verticalShootAxis,
                        PassKeyDown = _passKeyDown,
                        DashKeyDown = _dashKeyDown
                    }
                },
                States = new Entities.PlayerModule.States
                {
                    PlayerCarryPuck = _playerCarryPuckState,
                    PlayerNormal = _playerNormalState,
                    PlayerPass = _playerPassState,
                    PlayerDash = _playerDashState,
                    PlayerKnockedDown = _playerKnockedDownState,
                    PlayerSlapShot = _playerSlapShotState
                },
                StateMachine = new StateMachine<PlayerState>(_playerNormalState)
            };
            PlayerManager.Instance.Register(PlayerState);
        }

        public void Update()
        {
            PlayerState.StateMachine.Update(PlayerState);
        }

        public void FixedUpdate()
        {
            PlayerState.Player.UpdateRigidbody();
            PlayerState.StateMachine.FixedUpdate(PlayerState);
        }
    }
}

File: Assembly-FSharp/Assembly-FSharp/Views/PlayerView.fs
namespace Styx.Views

open UnityEngine

open Styx.Managers
open Styx.Entities.PlayerModule
open Styx.Models
open Styx.Commands
open RageCure.StateUtils
open Styx.States.Player
open Styx.Entities

type PlayerView () =
    inherit MonoBehaviour ()

    [<SerializeField>]
    let mutable playerId: PlayerId = Unchecked.defaultof<PlayerId>

    [<SerializeField>]
    let mutable rigidbodyModel: RigidbodyModel = Unchecked.defaultof<RigidbodyModel>
    [<SerializeField>]
    let mutable movementModel: MovementModel = Unchecked.defaultof<MovementModel>
    [<SerializeField>]
    let mutable animatorModel: AnimatorModel = Unchecked.defaultof<AnimatorModel>
    [<SerializeField>] 
    let mutable transformModel: TransformModel = Unchecked.defaultof<TransformModel>

    // *********************** INPUTS ***********************************
    [<SerializeField>]
    let mutable horizontalAxis: InputAxisCommand = Unchecked.defaultof<InputAxisCommand>
    [<SerializeField>]
    let mutable verticalAxis: InputAxisCommand =  Unchecked.defaultof<InputAxisCommand>
    [<SerializeField>]
    let mutable shootAxis: InputAxisCommand =  Unchecked.defaultof<InputAxisCommand>
    [<SerializeField>]
    let mutable passKeyDown: InputKeyDownCommand =  Unchecked.defaultof<InputKeyDownCommand>
    // *******************************************************************

    // *************************** STATES ******************************************
    [<SerializeField>]
    let mutable playerCarryPuckState: PlayerCarryPuckState =  Unchecked.defaultof<PlayerCarryPuckState>
    [<SerializeField>]
    let mutable playerNormalState: PlayerNormalState = Unchecked.defaultof<PlayerNormalState>
    [<SerializeField>]
    let mutable playerPassState: PlayerPassState = Unchecked.defaultof<PlayerPassState>
    [<SerializeField>]
    let mutable playerWristShotState: PlayerWristShotState = Unchecked.defaultof<PlayerWristShotState>
    // ******************************************************************************

    [<SerializeField>]
    let mutable opponentsGoal: Transform = null

    let mutable playerState: PlayerState = Unchecked.defaultof<PlayerState>

    member this.Start () =
        rigidbodyModel.Rigidbody <- this.GetComponent<Rigidbody>()
        animatorModel.Animator <- this.GetComponent<Animator>()
        transformModel.Transform <- this.GetComponent<Transform>()
        playerState <- 
            { player = 
                { id = playerId
                ; rigidbodyModel = rigidbodyModel
                ; movementModel = movementModel
                ; animatorModel = animatorModel
                ; transformModel = transformModel
                ; inputs = 
                    { horizontalAxis = horizontalAxis
                    ; verticalAxis = verticalAxis
                    ; shootAxis = shootAxis
                    ; passKeyDown = passKeyDown
                    }
                ; opponentsGoal = opponentsGoal
                }
            ; states = 
                { playerCarryPuck   = playerCarryPuckState
                ; playerNormal      = playerNormalState
                ; playerPass        = playerPassState
                ; playerWristShot   = playerWristShotState
                }
            ; stateMachine = new StateMachine<Player>(playerNormalState)
            }
        PlayerManager.Instance.Register playerState

    member this.Update () =
        playerState.stateMachine.Update(playerState.player)

    member this.FixedUpdate() =
        rigidbodyModel.Update()
        playerState.stateMachine.FixedUpdate(playerState.player)
        
    